### PR TITLE
Copy ASIO4ALL settings from old version

### LINKED ISSUE #1244: Installing new 64 bit Jamulus loses ASIO4ALL settings (& installer strongly recommends to remove the old version)
**Describe the bug**

The latest Jamulus builds for Windows are 64 bit, the previous versions were 32 bit. I think ASIO4ALL maintains customized settings per-exe file path and the install location is different for 32 vs 64 bit programs. This means the previously working settings for the old 32 bit version are not carried over to the new 64 bit version. Furthermore, the installer strongly recommends removing the older 32 bit version (this makes sense, because otherwise you end up with 2 sets of shortcuts, and things get kind of confusing).

**To Reproduce**

Have Jamulus working with ASIO4ALL on the old 32 bit version. Install a new 64 bit version, and click yes when installer strongly recommends removing old 32 bit version. Realize that ASIO4ALL is using the default settings now, and you don't remember what the good settings were. Reinstall the old Jamulus version to get the old settings, take a screenshot and get copy over the working settings to the 64 bit version.

I spent about an hour and a half talking someone through this over the phone yesterday.

**Expected behavior**

I don't think it's feasible for Jamulus to avoid this outright, but at the point where the installer recommends removing the old 32 bit Jamulus, it would be good to have a warning about it.

**Operating system**

This applies to Windows only.

**Jamulus version**

3.7.0, 64 bit builds.

### PR DESCRIPTION
**Pull request by ann0see**
Fixes #1244

### CODE REVIEW

File: windows/installer.nsi [hunk 1]
```diff
@@ -212,6 +218,30 @@ Section "Install_64Bit" INST_64
                 MessageBox MB_OKCANCEL|MB_ICONEXCLAMATION "$(OLD_VER_REMOVE_FAILED)" /sd IDCANCEL IDOK continueinstall
                 goto quit
             ${EndIf}
+
+            ; ask to move old ASIO4ALL registry configuration
+
+            ; check if old ASIO4ALL keys exit
+
+            ClearErrors
+            EnumRegKey $0 HKCU "SOFTWARE\ASIO4ALL v2 by Wuschel\8A9E7A56" 0
```
Comment (ann0see) [inline windows/installer.nsi:227]: Check if the key of the old version of A4A exists for the currently logged in user (not really a good way to check that. The currently logged in user might not have started Jamulus + A4A yet)

File: windows/installer.nsi [hunk 2]
```diff
@@ -212,6 +218,30 @@ Section "Install_64Bit" INST_64
...
+
+            IfErrors 0 continueinstall
+                MessageBox MB_YESNO|MB_ICONEXCLAMATION "Do you want to copy ASIO4ALL settings from the previous version if possible?" /sd IDYES IDNO continueinstall
```
Comment (ann0see) [inline windows/installer.nsi:230]: If we find this key, we'll ask if he wants to copy the settings (this message doesn't come up yet.)

File: windows/installer.nsi [hunk 3]
```diff
@@ -212,6 +218,30 @@ Section "Install_64Bit" INST_64
...
+
+                    IntOp $0 0 + 0
+                    EnumStart:
```
Comment (ann0see) [inline windows/installer.nsi:233]: For each user, replace the old with the new key contents

File: windows/installer.nsi [hunk 4]
```diff
@@ -212,6 +218,30 @@ Section "Install_64Bit" INST_64
                 MessageBox MB_OKCANCEL|MB_ICONEXCLAMATION "$(OLD_VER_REMOVE_FAILED)" /sd IDCANCEL IDOK continueinstall
                 goto quit
             ${EndIf}
```
Comment (ann0see) [inline windows/installer.nsi:220]: This code will only be executed if we find jamulus in the 32 bit prog. files folder and the current install is 64 bit.

**Review by ann0see — COMMENTED** [review]
Not yet ready; added some explanations

**Comment by ann0see** [comment]
Feel free to edit the code.

File: windows/installer.nsi (hunk 3 above)
Comment (softins) [reply windows/installer.nsi:233]: Should we do that? It feels wrong to update other users and not just HKCU.

Comment (ann0see) [reply windows/installer.nsi:233]: On my PC I have two users: 
1. Admin (user which I use to install Jamulus)
2. User (user which I use to use Jamulus)
If we don't  copy all users' settings, this issue would only be fixed for Admin but not for User.

**Comment by ann0see** [comment]
The latest commit makes it very simple. And seems work. We're not checking for anything, so probably it would also overwrite some other things.

File: windows/installer.nsi (hunk 3 above)
Comment (softins) [reply windows/installer.nsi]: But if Admin has never run Jamulus, it won't find the ASIO key in HKCU? Or if it does, the settings may well not be the same as were set the last time the User user Jamulus? Does the check for existing settings also need to scan the registry for users then?

The installer presumably has admin privileges, because I've never switched to admin to install Jamulus.

Comment (ann0see) [reply windows/installer.nsi]: > But if Admin has never run Jamulus, it won't find the ASIO key in HKCU? 

On my PC at least, Admin did run Jamulus. 

> Does the check for existing settings also need to scan the registry for users then?

Yes. This makes this code a lot more complicated.

Furthermore something like that would have to be re-implemented (but it's not that easy unfortunately):

foreach (user in users) {
    if (user.reg.oldkey.exists() && !user.reg.newkey.exists()) {
        user.reg.newkey.setValue(oldkey.getValue());
    }
}

Comment (hoffie) [reply windows/installer.nsi]: This pseudo-code sounds how it should work. Too bad that NSIS seems to be so complicated. :(

If NSIS implementation is the problem, would there be other workarounds? E.g. replacing the Jamulus shortcut by one pointing to a wrapper script which performs the per-user migration (maybe in Powershell?) and launches Jamulus.exe afterwards?
Not sure what side effects we would have to check for then or how it would look like to the user.

Comment (ann0see) [reply windows/installer.nsi]: NSIS should support that. We could use https://nsis.sourceforge.io/mediawiki/index.php?title=Registry_plug-in&oldid=16847 but I don't really want to include another plug-in. We already have too many in the installer.

**Comment by ann0see** [comment]
Tests so far with new code: 

1. Install new version on top of old version from non admin account with admin privileges from admin account: If I click "yes" in the my message, the settings get transfered. Note: the new registry key doesn't exist yet.
2. Remove new version and install old version --> old settings are saved for the old version
3. Install new version on top of old version after changes have been made in the new version. Press no to the copy message on the message in the installer --> Previously set settings from the new version get saved
4. Remove new version, install old version (without removing registry key), afterwards install new version. New key doesn't get touched. As expected. I'll need to move the message.
Still todo: 

1. Check what happens for the admin user (at least the new registry key seems to be created)

File: windows/deploy_windows.ps1 [hunk 5]
```diff
@@ -284,19 +284,19 @@ Function Build-NSProcess
...
         Remove-Item -Path "$WindowsPath\nsProcess\Release\" -Force -Recurse
         $OriginalEnv | % { Set-Item "Env:$($_.Name)" $_.Value }
     }
 }
 
-Clean-Build-Environment
-Install-Dependencies
-Build-App-Variants -QtInstallPath $QtInstallPath
-Build-NSProcess -QtInstallPath $QtInstallPath
+#Clean-Build-Environment
```
Comment (ann0see) [inline windows/deploy_windows.ps1:298]: This is used for debugging only!!!

File: windows/installer.nsi (hunk 3 above)
Comment (ann0see) [reply windows/installer.nsi]: Ok. I think I got it to work now.

File: windows/installer.nsi [hunk 6]
```diff
@@ -212,6 +218,34 @@ Section "Install_64Bit" INST_64
...
+
+            ; ask to move old ASIO4ALL registry configuration
+
+            MessageBox MB_YESNO|MB_ICONEXCLAMATION "If your're using ASIO4ALL: Do you want to copy ASIO4ALL settings from the previous version if possible? We will do this for every user." /sd IDYES IDNO continueinstall
```
Comment (ann0see) [inline windows/installer.nsi:224]: Do we need this message?

Comment (hoffie) [reply windows/installer.nsi:224]: I can't think of any reason why users would not want it.
As a user, I'd expect Jamulus+ASIO4ALL to work the same after the update than before. This logic tries to do that.
The logic is supposed to be non-destructive, i.e. it should not overwrite any configs (we should be really sure about that).

So my tendency would be to not add a (possibly confusing) message. Would also have the benefit of not needing another translators ping...

Comment (ann0see) [reply windows/installer.nsi]: Ok. Just did a force push without this change. I think now we'll just need to do some tests.

**Comment by ann0see** [comment]
Still one edge case: What should happen if an user now sets a different installation path? Do we need to handle this?

**Comment by hoffie** [comment]
> Still one edge case: What should happen if an user now sets a different installation path? Do we need to handle this?

That would have always been a problem, wouldn't it?

I think we could add a message on the page where the target path is chosen and warn about it ("If you change the installation path from previous installations, you may have to re-configure ASIO4ALL."). However, as this is not a regression as far as I understand) and because it would benefit from being translated, I'd suggest to postpone that part for 3.7.1?

**Comment by ann0see** [comment]
> That would have always been a problem, wouldn't it?

I think so.

**Review by softins — APPROVED** [review]
Appears to work ok on my W7 64-bit:
1. Installed 3.6.2
2. Made sure key 7A49ECC9 did not exist and 8A9E7A56 did exist.
3. Installed artifact from this PR. Was warned about removing the 32-bit version, which I accepted.
4. Checked registry and confirmed everything in 8A9E7A56 had been copied to 7A49ECC9.
5. Only one user on this PC had ever used Jamulus, so couldn't check for others.

**Comment by ann0see** [comment]
Logged in as admin: 
1. Removed new key in registry; installed the version from the artifact --> Previous settings are restored (since the new key gets created)

I think this is the most important check we need to make.

File: windows/installer.nsi [hunk 7]
```diff
@@ -212,6 +218,32 @@ Section "Install_64Bit" INST_64
...
+            EnumStart:
+                EnumRegKey $R1 HKEY_USERS "" $0 ; foreach user
+                IntOp $0 $0 + 1
+                StrCmp $R1 ".DEFAULT" EnumStart
+                StrCmp $R1 "" EnumEnd
+
+                ; check if new key already exists. If this is the case, we'll not continue
+                ClearErrors
+                EnumRegKey $1 HKU "$R1\SOFTWARE\ASIO4ALL v2 by Wuschel\7A49ECC9" 0
+                IfErrors 0 EnumStart ; if the above line gives an error, it can not find the key --> We'll continue
```
Comment (hoffie) [inline windows/installer.nsi:234]: Nit:
```suggestion
                IfErrors 0 EnumStart ; if the above line gives an error, it cannot find the key --> We'll continue
```

File: windows/installer.nsi [hunk 8]
```diff
@@ -212,6 +218,32 @@ Section "Install_64Bit" INST_64
...
+
+                ; check if old key already exists. If this is true, we'll continue and move the content of the old one to the new one.
```
Comment (hoffie) [inline windows/installer.nsi:236]:
```suggestion
                ; check if old key exists. If this is true, we'll continue and move the content of the old one to the new one.
```

File: windows/installer.nsi [hunk 9]
```diff
@@ -212,6 +218,32 @@ Section "Install_64Bit" INST_64
...
+                ClearErrors
+                EnumRegKey $1 HKU "$R1\SOFTWARE\ASIO4ALL v2 by Wuschel\8A9E7A56" 0
+                IfErrors EnumStart 0 ; if the above line gives an error, it can not find the key --> skip this user
```
Comment (hoffie) [inline windows/installer.nsi:239]:
```suggestion
                IfErrors EnumStart 0 ; if the above line gives an error, it cannot find the key --> skip this user
```

**Review by hoffie — APPROVED** [review]
I can't really comment on the NSH and NSI code as I have zero experience with it. I can't test on Windows either, sadly.

However, I'm confident that we should have this change and logic-wise it looks sounds.

I'd be happy if yet another person could test the installer, either explicitly or through another rc.

File: windows/NSISCopyRegistryKey/registry.nsh [hunk 10]
```diff
@@ -0,0 +1,166 @@
+!ifndef REGISTRY_NSH
```
Comment (pljones) [inline windows/NSISCopyRegistryKey/registry.nsh:1]: This code has no "watermark" -- it needs some declaration of authorship, licence, etc.

Comment (ann0see) [reply windows/NSISCopyRegistryKey/registry.nsh:1]: It should also be under the same license as nsProcess. Maybe we should just add it here?

Comment (ann0see) [reply windows/NSISCopyRegistryKey/registry.nsh:2]: Done now. I hope this comment is ok.

**Comment by ann0see** [comment]
Before merging this, I'd like to have feedback from @gilgongo

**Comment by gilgongo** [comment]
I can install it on my machine? It has old settings from a previous install. Where do I download it?

**Comment by ann0see** [comment]
I think the windows compilation process isn't finished yet, so you'll need to wait a bit. Afterwards you can find the binary on the checks page of this PR.

**Comment by ann0see** [comment]
I've uploaded the binary here: https://github.com/jamulussoftware/jamulus/releases/tag/untagged-964d405309c4e6a354c3

**Review by pljones — APPROVED** [review]

**Comment by ann0see** [comment]
I've merged this now since I think it works as expected.
